Parser/Token.cs

```diff
@@ -3,45 +3,51 @@ namespace PSI;
 
 // Represents a PSI language Token
 public class Token {
-   public Token (Tokenizer source, E kind, string text, int line, int column) 
-      => (Source, Kind, Text, Line, Column) = (source, kind, text, line, column);
+   public Token (Tokenizer source, E kind, string text, int line, int column, string msg)
+      => (Source, Kind, Text, Line, Column, ErrorMessage) = (source, kind, text, line, column, ErrorMessage + msg);
    public Tokenizer Source { get; }
    public E Kind { get; }
    public string Text { get; }
    public int Line { get; }
    public int Column { get; }
+   public string ErrorMessage { get; } = "";
 
    // The various types of token
    public enum E {
       // Keywords
       PROGRAM, VAR, IF, THEN, WHILE, ELSE, FOR, TO, DOWNTO,
       DO, BEGIN, END, PRINT, TYPE, NOT, OR, AND, MOD, _ENDKEYWORDS,
       // Operators
-      ADD, SUB, MUL, DIV, NEQ, LEQ, GEQ, EQ, LT, GT, ASSIGN, 
+      ADD, SUB, MUL, DIV, NEQ, LEQ, GEQ, EQ, LT, GT, ASSIGN,
       _ENDOPERATORS,
       // Punctuation
-      SEMI, PERIOD, COMMA, OPEN, CLOSE, COLON, 
+      SEMI, PERIOD, COMMA, OPEN, CLOSE, COLON,
       _ENDPUNCTUATION,
       // Others
       IDENT, INTEGER, REAL, BOOLEAN, STRING, CHAR, EOF, ERROR
    }
 
    // Print a Token
    public override string ToString () => Kind switch {
-      EOF or ERROR => Kind.ToString (),
+      EOF or ERROR => Text,
       < _ENDKEYWORDS => $"\u00ab{Kind.ToString ().ToLower ()}\u00bb",
       STRING => $"\"{Text}\"",
       CHAR => $"'{Text}'",
       _ => Text,
    };
 
    // Utility function used to echo an error to the console
-   public void PrintError () {
+   public void PrintError (bool isLast) {
       if (Kind != ERROR) throw new Exception ("PrintError called on a non-error token");
+      if (mY != Line) (mX, mY) = (0, Line);
+      mX = Math.Max (mX, Column - ErrorMessage.Length / 2) + 1 + mX;
       Console.ForegroundColor = ConsoleColor.Yellow;
-      Console.WriteLine ($"At line {Line}, column {Column} of {Source.FileName}: {Text}");
+      Console.WriteLine ("^", Console.CursorLeft = Column, Console.CursorTop += 1);
+      Console.WriteLine ($"{ErrorMessage}", Console.CursorLeft = mX);
       Console.ResetColor ();
+      Console.CursorTop = isLast ? Console.CursorTop - 1: Console.CursorTop - 3;

```

Comment:
Again, I think the requirement is that the PrintError (..) method should print the above 2 lines and the below 2 lines (if any) along with the error line. Here I see that only the error line along with the message is being printed by the above method. See the requirement below:
<img width="689" alt="image" src="https://github.com/Dakshita-Minocha/PSI-A18/assets/75829870/db512439-d581-45de-a4e0-593d28ca6634">

Reply:
I have that in Program.cs. Do you mean that I should shift all that code to PrintError?

Reply:
@Dakshita-Minocha : Yes. The main purpose of the above method is to print error message while compiling a given Pascal file. For the message to be more informative and elegant, the method should print 2 lines of content above and 2 lines below of the context apart from the actual error line.

Reply:
Okay, got it.